File: raspberry-pi-ios-client/RaspberryPiClient/Data/Infrastructure/Foxglove/Binary/FoxgloveBinaryMessageParser.swift
//
//  FoxgloveBinaryMessageParser.swift
//  RaspberryPiClient
//
//  Created by Shusuke Ota on 2025/03/20.
//

import Foundation

/// Foxglove バイナリメッセージのパーサ（サーバー → クライアント）
/// Ref: https://github.com/foxglove/ws-protocol/blob/main/docs/spec.md#binary-messages
struct FoxgloveBinaryMessageParser {
    /// サーバーから受信したバイナリデータをパースする
    /// - Parameter data: 受信したバイナリデータ（1バイト目が opcode）
    /// - Returns: パース結果
    static func parseServerMessage(_ data: Data) -> FoxgloveServerBinaryMessage {
        guard !data.isEmpty else {
            return .unknown(opcode: 0, error: .insufficientData)
        }

        let opcode = data[0]
        let payload = data.dropFirst(1)

        switch opcode {
        case FoxgloveServerBinaryOpcode.messageData.rawValue:
            return parseMessageData(Data(payload))
        case FoxgloveServerBinaryOpcode.serviceCallResponse.rawValue:
            return parseServiceCallResponse(Data(payload))
        default:
            return .unknown(opcode: opcode, error: nil)
        }
    }

    // MARK: - Message Data (0x01)
    // | 4 bytes: subscription id (uint32) |
    // | 8 bytes: timestamp (uint64) |
    // | remaining: payload |
    private static func parseMessageData(_ data: Data) -> FoxgloveServerBinaryMessage {
        guard data.count >= 12 else {
            return .unknown(opcode: 0x01, error: .insufficientData)
        }
        let subscriptionId = data.loadUInt32LittleEndian(at: 0)
        let timestamp = data.loadUInt64LittleEndian(at: 4)
        let payload = data.dropFirst(12)
        return .messageData(subscriptionId: subscriptionId, timestamp: timestamp, payload: Data(payload))
    }

    // MARK: - Service Call Response (0x03)
    // | 4 bytes: service id (uint32) |
    // | 4 bytes: call id (uint32) |
    // | 4 bytes: encoding length (uint32) |
    // | encoding length bytes: encoding (char[]) |
    // | remaining: payload |
    private static func parseServiceCallResponse(_ data: Data) -> FoxgloveServerBinaryMessage {
        guard data.count >= 12 else {
            return .unknown(opcode: 0x03, error: .insufficientData)
        }
        let serviceId = data.loadUInt32LittleEndian(at: 0)
        let callId = data.loadUInt32LittleEndian(at: 4)
        let encodingLength = Int(data.loadUInt32LittleEndian(at: 8))
        guard data.count >= 12 + encodingLength else {
            return .unknown(opcode: 0x03, error: .insufficientData)
        }
        let encodingData = data.dropFirst(12).prefix(encodingLength)
        guard let encoding = String(data: Data(encodingData), encoding: .utf8) else {
            return .unknown(opcode: 0x03, error: .invalidEncoding)
        }
        let payload = data.dropFirst(12 + encodingLength)
        return .serviceCallResponse(serviceId: serviceId, callId: callId, encoding: encoding, payload: Data(payload))
    }
}

/// サーバー → クライアントのバイナリメッセージ opcode（本ファイル内でのみ使用）
private enum FoxgloveServerBinaryOpcode: UInt8 {
    case messageData = 0x01
    case serviceCallResponse = 0x03
}

/// サーバーから受信したバイナリメッセージのパース結果
enum FoxgloveServerBinaryMessage {
    /// Message Data（opcode 0x01）
    /// subscription id, timestamp, payload
    case messageData(subscriptionId: UInt32, timestamp: UInt64, payload: Data)

    /// Service Call Response（opcode 0x03）
    /// service id, call id, encoding, payload
    case serviceCallResponse(serviceId: UInt32, callId: UInt32, encoding: String, payload: Data)

    /// 不明な opcode またはパースエラー
    case unknown(opcode: UInt8, error: FoxgloveBinaryParseError?)
}

enum FoxgloveBinaryParseError: Error {
    case insufficientData
    case invalidEncoding
}


private extension DataProtocol {
    /// アライメントに依存せず little-endian の UInt32 を読み取る（Data/Data.SubSequence 両対応、Index ベースで安全にアクセス）
    func loadUInt32LittleEndian(at offset: Int) -> UInt32 {
        let i0 = index(startIndex, offsetBy: offset)
        let i1 = index(startIndex, offsetBy: offset + 1)
        let i2 = index(startIndex, offsetBy: offset + 2)
        let i3 = index(startIndex, offsetBy: offset + 3)
        return UInt32(self[i0]) | (UInt32(self[i1]) << 8) | (UInt32(self[i2]) << 16) | (UInt32(self[i3]) << 24)
    }

    /// アライメントに依存せず little-endian の UInt64 を読み取る
    func loadUInt64LittleEndian(at offset: Int) -> UInt64 {
        UInt64(loadUInt32LittleEndian(at: offset)) | (UInt64(loadUInt32LittleEndian(at: offset + 4)) << 32)
    }
}

File: raspberry-pi-android-client/app/src/main/java/com/shusuke/raspberry_pi_client/data/infrastructure/foxglove_bridge/binary/FoxgloveBinaryMessageParser.kt
package com.shusuke.raspberry_pi_client.data.infrastructure.foxglove_bridge.binary

import java.nio.ByteBuffer
import java.nio.ByteOrder

/**
 * Foxglove バイナリメッセージのパーサ（サーバー → クライアント）
 *
 * Ref: https://github.com/foxglove/ws-protocol/blob/main/docs/spec.md#binary-messages
 */
object FoxgloveBinaryMessageParser {

    private const val OPCODE_MESSAGE_DATA: Byte = 0x01
    private const val OPCODE_SERVICE_CALL_RESPONSE: Byte = 0x03

    /**
     * サーバーから受信したバイナリデータをパースする
     *
     * @param data 受信したバイナリデータ（1バイト目が opcode）
     * @return パース結果
     */
    fun parseServerMessage(data: ByteArray): FoxgloveServerBinaryMessage {
        if (data.isEmpty()) {
            return FoxgloveServerBinaryMessage.Unknown(opcode = 0, error = FoxgloveBinaryParseError.InsufficientData)
        }
        val opcode = data[0]
        val payload = data.copyOfRange(1, data.size)
        return when (opcode) {
            OPCODE_MESSAGE_DATA -> parseMessageData(payload)
            OPCODE_SERVICE_CALL_RESPONSE -> parseServiceCallResponse(payload)
            else -> FoxgloveServerBinaryMessage.Unknown(opcode = opcode.toInt().and(0xFF).toByte(), error = null)
        }
    }

    /**
     * Message Data（opcode 0x01）
     * | 4 bytes: subscription id (uint32) |
     * | 8 bytes: timestamp (uint64) |
     * | remaining: payload |
     */
    private fun parseMessageData(data: ByteArray): FoxgloveServerBinaryMessage {
        if (data.size < 12) {
            return FoxgloveServerBinaryMessage.Unknown(opcode = OPCODE_MESSAGE_DATA, error = FoxgloveBinaryParseError.InsufficientData)
        }
        val buffer = ByteBuffer.wrap(data).order(ByteOrder.LITTLE_ENDIAN)
        val subscriptionId = buffer.int.toUInt() //and 0xFFFFFFFFu
        val timestamp = buffer.long
        val payload = data.copyOfRange(12, data.size)
        return FoxgloveServerBinaryMessage.MessageData(
            subscriptionId = subscriptionId,
            timestamp = timestamp,
            payload = payload,
        )
    }

    /**
     * Service Call Response（opcode 0x03）
     * | 4 bytes: service id (uint32) |
     * | 4 bytes: call id (uint32) |
     * | 4 bytes: encoding length (uint32) |
     * | encoding length bytes: encoding (char[]) |
     * | remaining: payload |
     */
    private fun parseServiceCallResponse(data: ByteArray): FoxgloveServerBinaryMessage {
        if (data.size < 12) {
            return FoxgloveServerBinaryMessage.Unknown(opcode = OPCODE_SERVICE_CALL_RESPONSE, error = FoxgloveBinaryParseError.InsufficientData)
        }
        val buffer = ByteBuffer.wrap(data).order(ByteOrder.LITTLE_ENDIAN)
        val serviceId = buffer.int.toUInt() and 0xFFFFFFFFu
        val callId = buffer.int.toUInt() and 0xFFFFFFFFu
        val encodingLength = buffer.int and 0x7FFFFFFF
        if (data.size < 12 + encodingLength) {
            return FoxgloveServerBinaryMessage.Unknown(opcode = OPCODE_SERVICE_CALL_RESPONSE, error = FoxgloveBinaryParseError.InsufficientData)
        }
        val encoding = data.copyOfRange(12, 12 + encodingLength).decodeToString()
        val payload = data.copyOfRange(12 + encodingLength, data.size)
        return FoxgloveServerBinaryMessage.ServiceCallResponse(
            serviceId = serviceId,
            callId = callId,
            encoding = encoding,
            payload = payload,
        )
    }
}

/**
 * サーバーから受信したバイナリメッセージのパース結果
 */
sealed class FoxgloveServerBinaryMessage {
    data class MessageData(
        val subscriptionId: UInt,
        val timestamp: Long,
        val payload: ByteArray,
    ) : FoxgloveServerBinaryMessage() {
        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (javaClass != other?.javaClass) return false
            other as MessageData
            if (subscriptionId != other.subscriptionId) return false
            if (timestamp != other.timestamp) return false
            return payload.contentEquals(other.payload)
        }

        override fun hashCode(): Int {
            var result = subscriptionId.hashCode()
            result = 31 * result + timestamp.hashCode()
            result = 31 * result + payload.contentHashCode()
            return result
        }
    }

    data class ServiceCallResponse(
        val serviceId: UInt,
        val callId: UInt,
        val encoding: String,
        val payload: ByteArray,
    ) : FoxgloveServerBinaryMessage() {
        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (javaClass != other?.javaClass) return false
            other as ServiceCallResponse
            if (serviceId != other.serviceId) return false
            if (callId != other.callId) return false
            if (encoding != other.encoding) return false
            return payload.contentEquals(other.payload)
        }

        override fun hashCode(): Int {
            var result = serviceId.hashCode()
            result = 31 * result + callId.hashCode()
            result = 31 * result + encoding.hashCode()
            result = 31 * result + payload.contentHashCode()
            return result
        }
    }

    data class Unknown(
        val opcode: Byte,
        val error: FoxgloveBinaryParseError?,
    ) : FoxgloveServerBinaryMessage()
}

enum class FoxgloveBinaryParseError {
    InsufficientData,
    InvalidEncoding,
}
